File: 2458-height-of-binary-tree-after-subtree-removal-queries/2458-height-of-binary-tree-after-subtree-removal-queries.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int solve(TreeNode* root, int lev, vector<int>& level, vector<int>& height, vector<int>& maxLevelHeight, vector<int>& secondMaxLevelHeight){
        if(root == NULL) return 0;

        level[root->val] = lev;

        int l = solve(root->left, lev + 1, level, height, maxLevelHeight, secondMaxLevelHeight);
        int r = solve(root->right, lev + 1, level, height, maxLevelHeight, secondMaxLevelHeight);

        height[root->val] = 1 + max(l, r);

        if(maxLevelHeight[lev] < height[root->val]){
            secondMaxLevelHeight[lev] = maxLevelHeight[lev];
            maxLevelHeight[lev] = height[root->val];
        }
        else if(secondMaxLevelHeight[lev] < height[root->val]){
            secondMaxLevelHeight[lev] = height[root->val];
        }

        return height[root->val];
    }
public:
    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        vector<int> level(1e5+1, 0);
        vector<int> height(1e5+1, 0);
        vector<int> maxLevelHeight(1e5+1, 0);
        vector<int> secondMaxLevelHeight(1e5+1, 0);
        vector<int> ans(queries.size(), 0);

        solve(root, 0, level, height, maxLevelHeight, secondMaxLevelHeight);

        for(int i = 0; i < queries.size(); i++){
            int query = queries[i];
            int lev = level[query];
            int maxHeight = maxLevelHeight[lev];
            int secondMaxHeight = secondMaxLevelHeight[lev];

            if(maxHeight == height[query]){
                ans[i] = lev + secondMaxHeight - 1;
            }
            else{
                ans[i] = lev + maxHeight - 1;
            }
        }

        return ans;
    }
};

File: 2458-height-of-binary-tree-after-subtree-removal-queries/2458-height-of-binary-tree-after-subtree-removal-queries.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int solve(TreeNode root, int lev, int[] height, int[] level, int[] levelMaxHt, int[] secLevelMaxHt){
        if(root == null) return 0;

        level[root.val] = lev;

        int l = solve(root.left, lev+1, height, level, levelMaxHt, secLevelMaxHt);
        int r = solve(root.right, lev+1, height, level, levelMaxHt, secLevelMaxHt);

        height[root.val] = 1 + Math.max(l, r);

        if(levelMaxHt[lev] < height[root.val]){
            secLevelMaxHt[lev] = levelMaxHt[lev];
            levelMaxHt[lev] = height[root.val]; 
        }
        else if(secLevelMaxHt[lev] < height[root.val]) secLevelMaxHt[lev] = height[root.val];

        return height[root.val];
    }

    public int[] treeQueries(TreeNode root, int[] queries) {
        int[] height = new int[100001];
        int[] level = new int[100001];
        int[] levelMaxHt = new int[100001];
        int[] secLevelMaxHt = new int[100001];
        int n = queries.length;
        int[] ans = new int[n];

        solve(root, 0, height, level, levelMaxHt, secLevelMaxHt);

        for(int i = 0; i < n; i++){
            int query = queries[i];
            int lev = level[query];
            int ht = height[query];
            int maxHt = levelMaxHt[lev];
            int secMaxHt = secLevelMaxHt[lev];

            if(maxHt == ht) ans[i] = lev + secMaxHt - 1;
            else ans[i] = lev + maxHt - 1;
        }

        return ans;
    }
}